File: frontend/src/services/authService.ts
import api from './api'
import type { AuthResponse, LoginRequest, RegisterRequest } from '../types/Auth'

const register = async (
  data: RegisterRequest
): Promise<AuthResponse> => {
  const response = await api.post<AuthResponse>(
    '/api/auth/register',
    data
  )

  return response.data
}

const login = async (
  data: LoginRequest
): Promise<AuthResponse> => {
  const response = await api.post<AuthResponse>(
    '/api/auth/login',
    data
  )

  return response.data
}

export const authService = {
  register,
  login,
}

File: backend/Services/AuthService.cs
using backend.Data;
using backend.Dtos.Auth;
using backend.Models;
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;

namespace backend.Services;

public class AuthService
{
    private readonly AppDbContext _context;
    private readonly PasswordHasher<User> _passwordHasher;

    public AuthService(AppDbContext context)
    {
        _context = context;
        _passwordHasher = new PasswordHasher<User>();
    }

    public async Task<AuthResponse> RegisterAsync(RegisterRequest request)
    {
        var existingUser = await _context.Users
            .FirstOrDefaultAsync(user => user.Username == request.Username);

        if (existingUser is not null)
        {
            throw new InvalidOperationException("Username already exists.");
        }

        var user = new User
        {
            FullName = request.FullName,
            Username = request.Username
        };

        user.PasswordHash = _passwordHasher.HashPassword(user, request.Password);

        _context.Users.Add(user);
        await _context.SaveChangesAsync();

        return new AuthResponse
        {
            Token = string.Empty,
            Username = user.Username
        };
    }

    public async Task<AuthResponse?> LoginAsync(LoginRequest request)
    {
        var user = await _context.Users
            .FirstOrDefaultAsync(user => user.Username == request.Username);

        if (user is null)
        {
            return null;
        }

        var result = _passwordHasher.VerifyHashedPassword(
            user,
            user.PasswordHash,
            request.Password);

        if (result == PasswordVerificationResult.Failed)
        {
            return null;
        }

        return new AuthResponse
        {
            Token = "test-token",
            Username = user.Username
        };
    }
}